agent/dns.go

```diff
@@ -1483,78 +1372,308 @@ func findWeight(node structs.CheckServiceNode) int {
+			&dns.SRV{
+				Hdr: dns.RR_Header{
+					Name:   q.Name,
+					Rrtype: dns.TypeSRV,
+					Class:  dns.ClassINET,
+					Ttl:    uint32(ttl / time.Second),
+				},
+				Priority: 1,
+				Weight:   uint16(findWeight(serviceNode)),
+				Port:     uint16(d.agent.TranslateServicePort(dc, serviceNode.Service.Port, serviceNode.Service.TaggedAddresses)),
+				Target:   ipFQDN,
+			},
+		}
+
+		ipRecord.Header().Name = ipFQDN
+		return answers, []dns.RR{ipRecord}
+	}
+
+	ipRecord.Header().Name = q.Name
+	return []dns.RR{ipRecord}, nil
+}
+
+// Craft dns records for an FQDN
+// In case of an SRV query the answer will be a IN SRV and additional data will store an IN A to the IP
+// Otherwise it will return a CNAME and a IN A record
+func (d *DNSServer) makeRecordFromFQDN(dc string, fqdn string, serviceNode structs.CheckServiceNode, req *dns.Msg, ttl time.Duration, cfg *dnsConfig, maxRecursionLevel int) ([]dns.RR, []dns.RR) {
+	edns := req.IsEdns0() != nil
+	q := req.Question[0]
+
+	more := d.resolveCNAME(cfg, dns.Fqdn(fqdn), maxRecursionLevel)
+	var additional []dns.RR
+	extra := 0
+MORE_REC:
+	for _, rr := range more {
+		switch rr.Header().Rrtype {
+		case dns.TypeCNAME, dns.TypeA, dns.TypeAAAA:
+			// set the TTL manually
+			rr.Header().Ttl = uint32(ttl / time.Second)
+			additional = append(additional, rr)
+
+			extra++
+			if extra == maxRecurseRecords && !edns {
+				break MORE_REC
+			}
+		}
+	}
+
+	if q.Qtype == dns.TypeSRV {
+		answers := []dns.RR{
+			&dns.SRV{
+				Hdr: dns.RR_Header{
+					Name:   q.Name,
+					Rrtype: dns.TypeSRV,
+					Class:  dns.ClassINET,
+					Ttl:    uint32(ttl / time.Second),
+				},
+				Priority: 1,
+				Weight:   uint16(findWeight(serviceNode)),
+				Port:     uint16(d.agent.TranslateServicePort(dc, serviceNode.Service.Port, serviceNode.Service.TaggedAddresses)),
+				Target:   dns.Fqdn(fqdn),
+			},
+		}
+		return answers, additional
+	}
+
+	answers := []dns.RR{
+		&dns.CNAME{
+			Hdr: dns.RR_Header{
+				Name:   q.Name,
+				Rrtype: dns.TypeCNAME,
+				Class:  dns.ClassINET,
+				Ttl:    uint32(ttl / time.Second),
+			},
+			Target: dns.Fqdn(fqdn),
+		}}
+	answers = append(answers, additional...)
+
+	return answers, nil
+}
+
+func (d *DNSServer) nodeServiceRecords(dc string, node structs.CheckServiceNode, req *dns.Msg, ttl time.Duration, cfg *dnsConfig, maxRecursionLevel int) ([]dns.RR, []dns.RR) {
+	serviceAddr := d.agent.TranslateServiceAddress(dc, node.Service.Address, node.Service.TaggedAddresses)
+	nodeAddr := d.agent.TranslateAddress(node.Node.Datacenter, node.Node.Address, node.Node.TaggedAddresses)
+
+	nodeIPAddr := net.ParseIP(nodeAddr)
+	serviceIPAddr := net.ParseIP(serviceAddr)
+
+	// There is no service address and the node address is an IP
+	if serviceAddr == "" && nodeIPAddr != nil {
+		//if node.Node.Datacenter != cfg.Datacenter {
+		if node.Node.Address != nodeAddr {
+			// Do not CNAME node address in case of WAN address
+			return d.makeRecordFromIP(dc, nodeIPAddr, node, req, ttl)
+		}
```

Comment:
I don't quite understand the purpose of this conditional. The comment mentions CNAME RRs for WAN addresses but `makeRecordFromServiceNode` seems to only generate A/AAAA or SRV RRs.

The main difference I see in `makeRecordFromIP` and `makeRecordFromServiceNode` is that the SRV Target generated is a node name in one and the odd encoded IP form in another. For wan translated addresses we are returning the encoded ip form.

So I guess my question is why we need to use that encoded ip form at all here? Why would returning the record with a Target of `<node name>.node.<dc>.<domain>` not work?

Reply:
Originaly it was not needed as we return directly a A to the IP of the service but
this has been done to comply to test TestDNS_ServiceLookup_WanTranslation/node-wan-from-dc1 which expects generated name:
```expected: "7f000002.addr.dc2.consul."```
```actual  : "foo.node.dc2.consul."```

With this conditional, we return
```7f000002.addr.dc2.consul. IN A 127.0.0.2```
Without:
```foo.node.dc2.consul. IN A 127.0.0.2```

Reply:
I guess that makes sense. It will prevent future RPCs because we can resolve the ip encoded address without any other knowledge.